### DISCUSSION TITLE
DAI 2023-2024 - Java IOs benchmarking - Thomas Vuilleumier

### DISCUSSION THREAD

**Discussion by Cobora2001** [heig-vd-dai-course/heig-vd-dai-course#101] [Show and tell]
# Java IOs benchmarking

## Repository

[My repository is here !](https://github.com/Cobora2001/heig-vd-dai-course-java-ios-practical-content)

### Between buffered or not

**Comparing : Binary Buffered and not Buffered in Writing**
|Buffered ?|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Yes| 0ms / 731700ns | 0ms / 620600ns | 10ms / 10386400ns | 95ms / 95377500ns |
|No| 0ms / 398200ns | 3ms / 3176900ns | 2146ms / 2146755200ns | 10861ms / 10861482300ns |

**Comparing : Binary Buffered and not Buffered in Reading**
|Buffered ?|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Yes| 0ms / 258800ns | 0ms / 238600ns | 20ms / 20951500ns | 93ms / 93186700ns |
|No| 0ms / 142600ns | 2ms / 2777300ns | 1183ms / 1183049400ns | 6033ms / 6033020500ns |

**Comparing : Text Buffered and not Buffered in Writing**
|Buffered ?|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Yes| 0ms / 757100ns| 2ms / 2027500ns | 16ms / 16969800ns | 121ms / 121283500ns |
|No| 0ms / 492800ns | 1ms / 1987200ns | 54ms / 54213000ns | 123ms / 123299500ns |

**Comparing : Text Buffered and not Buffered in Reading**
|Buffered ?|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Yes| 0ms / 549500ns | 0ms / 335600ns | 24ms / 24351300ns | 96ms / 96869100ns |
|No| 0ms / 156600ns | 1ms / 1006600ns | 39ms / 39114500ns | 117ms / 117402500ns |

### Between binary and text

**Comparing : Buffered in Writing**
|Type|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Binary | 0ms / 731700ns | 0ms /  620600ns | 10ms / 10386400ns |   95ms /   95377500ns |
|Text    | 0ms / 757100ns | 2ms / 2027500ns | 16ms / 16969800ns | 121ms / 121283500ns |

**Comparing : Buffered in Reading**
|Type|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Binary| 0ms / 258800ns | 0ms / 238600ns | 20ms / 20951500ns | 93ms / 93186700ns |
|Text    | 0ms / 549500ns | 0ms / 335600ns | 24ms / 24351300ns | 96ms / 96869100ns |

**Comparing : Not Buffered in Writing**
|Type|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Binary| 0ms / 398200ns | 3ms / 3176900ns | 2146ms / 2146755200ns | 10861ms / 10861482300ns |
|Text    | 0ms / 492800ns | 1ms / 1987200ns |    54ms /     54213000ns |     123ms /     123299500ns |

**Comparing : Not Buffered in Reading**
|Type|1 byte|1'024 bytes|1'048'576 bytes|5'242'880 bytes|
|-----------|-------|------------|----------------|----------------|
|Binary| 0ms / 142600ns | 2ms / 2777300ns | 1183ms / 1183049400ns | 6033ms / 6033020500ns |
|Text    | 0ms / 156600ns | 1ms / 1006600ns |    39ms /      39114500ns |  117ms /   117402500ns |

## Conclusions

Notice : These results were obtained on one computer during one test. As such, they may not be entirely representative and shouldn't be taken as universal truth. They should be able to show us trends though, which we will discuss here.

#### Binary

For a single byte, it is generally better to have direct access instead of buffered access, for both read and write, but that is the only case where this is so. From 1'024 bytes and on, having a binary file being written or read via a buffer is more and more effective as we read and write more and more bytes (in our example, getting to 100x more effective when writing 5'242'880 bytes).

#### Text

Similar to binary files, text files are faster to either read and write when they have one byte when not buffered. However, the story is more complex from there on, and the differences are less remarkable.

When reading more than one byte, it is generally useful to have a buffered reader. But the difference is relatively minor, only going approximately 1,5x to 3.5x faster with a buffer, depending on the file size, which is far from the 50x faster when reading binary files in some instances.

Regarding writing, this is more complicated. For 1024 bytes, it was still slower to have a buffer. Then, for 1'048'576 bytes, it was faster by a few times, but then, with 5'242'880 bytes, it was only just faster when buffered. This seems to be far from a linear relationship

### Why is it more efficient than the other types of streams?

For a single byte, creating a buffer is note useful, which explains these results.

Regarding the binary files, it is quite logical to obtain better results with a buffer, given that this prevents having to access the file each time, and that it is only writing 0 or 1 to the buffer.

In contrast, with text files, it isn't only 0 or 1, but any characters available in the encoding set, meaning that the "translation" needs to happen for every character, buffer or not. This dampens the difference gained by not having to access the file. For the more detailed explanations, I do not know.

### What is the difference between binary data and text data?

Both are represented via 0 and 1, but the text file from there needs to be interpreted via an encoding, which also takes some time and will affect our results.

### What is a character encoding?

When representing a character, one needs to know how to write it. In a world of 0 and 1, multiple characters will in reality be needed to write anything more complex than a binary result. We need to be able to translate a series of 0 and 1 into (a) character(s), and for that, we need a translation dictionary.

This dictionary is a character set. There are multiple character sets, given that there are a lot of characters to write, but some will only be needed in some cases, making it useless and wasteful to have in a specific character set that won't have to deal with this case. The more characters you want to represent, the more bits you will need to represent a single character. Having the right character set for you case means lighter files, and more free space.

It does induce some difficulties. For example, if you get a file without knowing how it was encoded, you won't be able to read (parts of) it (or at least you will need some research to find the initial encoding).

### Why is this methodology important?

Because empirical knowledge, if done well, is a great source of insight for the scientific method. To have insight about the real world, we need to be able to gather data from it.

To make rules to know what will happen, we look first at what does happen in the real world, and then we try and suppose what the rules are based on the data. We then test the rules we have against new data by trying to do predictions, and if it works, we have a useful rule, that will be replaced when a better rule, meaning a rule that correlates better to the data from the real world (often new data), is discovered. Everything is based on quitting only thinking in our minds, and gathering data to apply this thinking to.

Even philosophy, which was the ultimate art of the discovery by thought alone (since its separation from natural science), has started having inputs from polls and other methods of data gathering, trying to tie the theories that they were gathered to the real data. This is a step we can't ignore if we are to try to reach conclusions.

(Note : As stated earlier, my method was insufficient, given I only did one test. But this was, I hope, good enough to reach some general conclusions, instead of precise ones)

**Comment by ludelafo** [comment]
I am really sorry, I was going through the GitHub Discussions and only noticed now that your discussion was never answered.

Your explanations are all good, just be aware that you should specify the encoding of your text files when reading/writing them in your code. I see you imported the right imports (`import java.nio.charset.StandardCharsets;`) but you don't make usage of them.
